bokehjs/src/coffee/models/sources/column_data_source.coffee

```diff
@@ -49,13 +50,58 @@ export stream_to_column = (col, new_col, rollover) ->
   return concat_typed_arrays(col, tmp)
 
 # exported for testing
-export patch_to_column = (col, patch) ->
-  for i in [0...patch.length]
-    [ind, value] = patch[i]
-    col[ind] = value
+export slice = (ind, length) ->
+  if isObject(ind)
+    return [ind.start ? 0, ind.stop ? length, ind.step ? 1]
+  return [start, stop, step] = [ind, ind+1, 1]
 
+# exported for testing
+export patch_to_column = (col, patch, shapes) ->
```

Comment:
@bryevdv, I will need a signature for this function. My rough approximation is this:
```
(col: any[],
 patch: [number | [number, number] | [number, number, number], any][],
 shape: [number, number][]) => Set<number>
```
btw. `[type_0, ..., type_n]` syntax stands for a tuple (on the type level, because this isn't (easily) distinguishable at runtime).